Moving the HandConstraint Solver from Experimental to Release

Cameron-Micka opened the pull request:
## Overview
This pull request moves the HandConstraint  (and HandConstraintPalmUp) solver from Experimental to Release by adding tests, documentation, and a handful of bug fixes and improvements. Functionality remains pretty much the same from the previous pull request (https://github.com/microsoft/MixedRealityToolkit-Unity/pull/4532) and utilizes solver fixes from @Troy-Ferrell's (https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5416)

The solver has also be genericized to work with motion controllers.

Thank you @julenka for guidance on how to write unit tests that utilize hands!

## Changes
- Fixes: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5312, https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5304, https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5257, https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5413


## Verification
Please try HandBasedMenuExample scene.
>
> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

Troy-Ferrell commented on Documentation/InputSystem/HandTracking.md:
```diff
@@ -161,3 +161,41 @@ public class MyHandMeshEventHandler : IMixedRealityHandMeshHandler
+
+There is currently a known issue with Master builds using the .NET backend. In .NET Native, `IInspectable` pointers cannot be marshaled from native to managed code using `Marshal.GetObjectForIUnknown`. The MRTK uses this to obtain the `SpatialCoordinateSystem` in order to receive hand and eye data from the platform.
+
+We've provided DLL source as a workaround for this issue, in [the native Mixed Reality Toolkit repo](https://github.com/microsoft/MixedRealityToolkit/tree/master/DotNetNativeWorkaround). Please follow the instructions in the README there and copy the resulting binaries into a Plugins folder in your Unity assets. After that, the WindowsMixedRealityUtilities script provided in the MRTK will resolve ...
```
We've provided a DLL and C# wrapper/utility/source as a workaround for this issue

Does WindowsMixedRealityUtilities show up as xref?

Cameron-Micka replied:
To be honest I have no idea where this came from haha. @keveleigh is this because this PR https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5458/files is not in stabilization?

Troy-Ferrell commented on Documentation/README_Solver.md:
```diff
@@ -22,6 +22,8 @@ The third category is the solver itself. The following solvers provide the build
 * **SurfaceMagnetism**, casts rays to surfaces in the world, and align the object to that surface.
 * **Momentum**: Applies acceleration/velocity/friction to simulate momentum and springiness for an object being moved by other solvers/components.
 * **InBetween**: Keeps an object in between two tracked objects.
+* **HandConstraint**: Keeps an object in a region safe for hand constrained interactive content.
```
Contrains object to follow hands in a region that doesn't intersect the GameObject with the hands. Useful for hand constrained interactive content such as menus etc.

region safe is a big question mark until I read the section below

Troy-Ferrell commented on Documentation/README_Solver.md:
```diff
@@ -22,6 +22,8 @@ The third category is the solver itself. The following solvers provide the build
 * **Momentum**: Applies acceleration/velocity/friction to simulate momentum and springiness for an object being moved by other solvers/components.
 * **InBetween**: Keeps an object in between two tracked objects.
+* **HandConstraint**: Keeps an object in a region safe for hand constrained interactive content.
+* **HandConstraintPalmUp**: Derives from HandConstraint, also includes logic to test if the palm is facing the user before activation.
```
nit: Derives from HandConstraint but includes logic...

Troy-Ferrell commented on Documentation/README_Solver.md:
```diff
@@ -36,5 +38,19 @@ When *UpdateLinkedTransform* is true, the solver will calculate position & orien
 
+## Hand constraint solvers ##
+
+The **HandConstraint** behavior provides a solver that constrains the tracked object to a region safe for hand constrained content (such as hand UI, menus, etc). Safe regions are considered areas that don't intersect with the hand. A derived class of **HandConstraint** called **HandConstraintPalmUp** is also included to demonstrate a common behavior of activating the solver tracked object when th ...
```
Please xref HandConstraint & HandConstraintPalmUp.

FYI, @Cameron-Micka  I am working on a change that will update the SolverExample scenes and this documentation. The change are all for the above general solver information

Troy-Ferrell commented on Documentation/README_Solver.md:
```diff
@@ -36,5 +38,19 @@ When *UpdateLinkedTransform* is true, the solver will calculate position & orien
+
+<img src="../Documentation/Images/Solver/MRTK_Solver_HandConstraintSafeZones.PNG" width="450">
+
+* **Activation Events**: Currently the **HandConstraint** triggers four events, OnHandActivate, OnHandDecativate, OnFirstHandDetected, and OnLastHandDetected. The OnHandActivate events triggers when a hand satisfies the IsHandActive method and OnHandDecativate triggers when the IsHandActive method is no longer satisfied. OnFirstHandDetected occurs when state changes from no hands in view, to the  ...
```
maybe just breakout each event with description instead of one long paragraph

* **ActivationEvents**
- OnHandActivate: bla blah blah
- OnHandDeactivate: blah blah blah

julenka commented:
I'm out on vacation until 8/7, will have a look when I get back on 8/8

Troy-Ferrell commented on Documentation/InputSystem/HandTracking.md:
```diff
@@ -13,7 +13,7 @@ Joint prefabs are visualized using simple prefabs. The _Palm_ and _Index Finger_
 
 > [!WARNING]
-> Avoid using complex scripts or expensive rendering in joint prefabs, since joint objects are transformed on every frame and can have significant performance cost!**
+> Avoid using complex scripts or expensive rendering in joint prefabs, since joint objects are transformed on every frame and can have significant performance cost!
```
nit: remove ! at end

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/SolverTests.cs:
```diff
@@ -15,14 +15,16 @@
 namespace Microsoft.MixedReality.Toolkit.Tests
 {
-    public class SolverTests : BasePlayModeTests
+    public class SolverTests
```
instead of breaking inheritance, can you just add virtual to the BasePlayModeTests function and then override here?

Troy-Ferrell commented on Assets/MixedRealityToolkit.Tests/PlayModeTests/SolverTests.cs:
```diff
@@ -35,6 +37,25 @@ protected class SetupData
+        [TearDown]
+        public void TearDown()
+        {
+            foreach (var setupData in setupDataList)
```
Good catch! I didn't know the scene wasn't wiped between tests

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/Solver.cs:
```diff
@@ -80,7 +80,12 @@ protected Quaternion GoalRotation
             get { return SolverHandler.GoalScale; }
-            set { SolverHandler.GoalScale = value; }
+            set {
+                if (SolverHandler == null)
```
If we are going to go this path of searching fo solver handler. Can you move to util method and call for Awake, OnEnable, GoalPosition, GoalRotation, etc?

Cameron-Micka replied:
I'm actually not sure why we needs these? Looks like they were added here: https://github.com/microsoft/MixedRealityToolkit-Unity/commit/b042ac1446aecc4d5190501620a036d868fb064d#diff-8b68ce5f0aa5c4660f69b67600085bed I'll remove them since it doesn't look like they are required. (Since we have [RequireComponent(typeof(SolverHandler))] and acquire it on Awake)

Troy-Ferrell replied:
I think we still need them on Awake

Troy-Ferrell commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraint.cs:
```diff
@@ -178,28 +193,18 @@ protected IMixedRealityInputSystem InputSystem
                 if (inputSystem == null)
                 {
-                    MixedRealityServiceRegistry.TryGetService<IMixedRealityInputSystem>(out inputSystem);
+                    MixedRealityServiceRegistry.TryGetService(out inputSystem);
```
Why remove the Interface?

Cameron-Micka replied:
Just for brevity since it can be inferred via the input.

Troy-Ferrell replied:
huh, and it still works? If no interface is provided then it will be IMixedRealityService which shouldn't be a valid key

Cameron-Micka replied:
AFAIK it works since C# will infer generic types based on arguments. Just checked the debugger to make sure:
![image](https://user-images.githubusercontent.com/13305729/62401524-6cd1de00-b538-11e9-9c6e-61f106fd8258.png)

Troy-Ferrell replied:
Ahhh because of the input parameter. I ran into an issue with this but there was no T serviceInstance to infer from

Troy-Ferrell commented on Documentation/README_Solver.md:
```diff
@@ -36,5 +38,19 @@ When *UpdateLinkedTransform* is true, the solver will calculate position & orien
+<img src="../Documentation/Images/Solver/MRTK_Solver_HandConstraintSafeZones.PNG" width="450">
+
+* **Activation Events**: Currently the **HandConstraint** triggers four events, OnHandActivate, OnHandDecativate, OnFirstHandDetected, and OnLastHandDetected. The OnHandActivate events triggers when a hand satisfies the IsHandActive method and OnHandDecativate triggers when the IsHandActive method is no longer satisfied. OnFirstHandDetected occurs when state changes from no hands in view, to the  ...
+
```
Nit: Do you think we should mention the requirement of HandsBounds? or just leave it be since ideally developers don't even need to reference it or mess with it

Cameron-Micka replied:
I could go either way, they kind of exist behind the scenes aren't dependent on the Solver system. So maybe they could be documented elsewhere?

Cameron-Micka commented:
> So awesome!
> Can you also update the PR description to include the key information about the .NET workaround?

Thank you for the reveiw! I think I addressed most of your comments. I just need to figure out how to address changes that aren't really mine. @wiwei I made this PR before the stabilization branch was cut (it was built of development). Should I target development or stabilization?

wiwei commented:
@Cameron-Micka I think that we'll probably wait for tomorrow/Thursday for @julenka to take a look at this one per her comments above

Cameron-Micka commented:
> @Cameron-Micka I think that we'll probably wait for tomorrow/Thursday for @julenka to take a look at this one per her comments above

Sounds good, thank you for the update!

julenka commented:
/azp run

julenka commented:
> MixedRealityToolkit.Examples/Demos/HandTracking/Prefabs/UlnarConstrainedHandMenu.prefab 

I had to look up Ulnar :-) Any reason you renamed it from palm (much easier to remember), other than being precise?

Cameron-Micka commented on Assets/MixedRealityToolkit.SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        [SerializeField]
+        [Tooltip("Do the fingers on the hand need to be straitened, rather than curled, to form a flat hand shape.")]
```
😂

Cameron-Micka commented:
/azp run

Cameron-Micka commented:
> Overall, looks good! I had some comments regarding motion controller support -- we should mention in docs + comments where motion controllers are supported (and not), as well as behavior of different follow modes when using motion controllers.

Thank you for the review! I've made a bunch of documentation updates to point out when motion controllers work.

Cameron-Micka commented:
/azp run

julenka commented on Documentation/README_Solver.md:
```diff
@@ -23,6 +23,8 @@ The third category is the solver itself. The following solvers provide the build
 * [`Momentum`](xref:Microsoft.MixedReality.Toolkit.Utilities.Solvers.Momentum): Applies acceleration/velocity/friction to simulate momentum and springiness for an object being moved by other solvers/components.
 * [`InBetween`](#inbetween): Keeps an object in between two tracked objects.
+* [`HandConstraint`](#handconstraint-and-handconstraintpalmup): Constrains object to follow hands in a region that doesn't intersect the GameObject with the hands. Useful for hand constrained interactive content such as menus, etc. This solver is intended to work with [IMixedRealityHand](xref:Microsoft.MixedReality.Toolkit.Input.IMixedRealityHand) but also works with [IMixedRealityController](xre ...
+* [`HandConstraintPalmUp`](#handconstraint-and-handconstraintpalmup): Derives from HandConstraint but includes logic to test if the palm is facing the user before activation. This solver only works with [IMixedRealityHand](xref:Microsoft.MixedReality.Toolkit.Input.IMixedRealityHand) controllers, with other controller types this solver will behave just like it's base class.
```
```suggestion
* [`HandConstraintPalmUp`](#handconstraint-and-handconstraintpalmup): Derives from HandConstraint but includes logic to test if the palm is facing the user before activation. This solver only works with [IMixedRealityHand](xref:Microsoft.MixedReality.Toolkit.Input.IMixedRealityHand) controllers, with other controller types this solver will behave just like its base class.
```

julenka commented:
All tests passed, then I fixed a typo (it's => its), which resets the test status. For something as small as a typo fix I'm willing to take the risk, since I think it's extremely unlikely that this will break mrtk_docs PR. Completing the PR so we can move on to other bug fixes / validation or GA release.
